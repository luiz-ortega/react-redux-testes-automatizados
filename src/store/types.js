export const FETCH_PLANETS = 'FETCH_PLANETS';
export const FETCH_PLANETS_SUCCESS = 'FETCH_PLANETS_SUCCESS';
export const FETCH_PLANETS_FAILURE = 'FETCH_PLANETS_FAILURE';
export const SORT_PLANET_LIST = 'SORT_PLANET_LIST';
export const FILTER_PLANET_LIST_BY_NAME = 'FILTER_PLANET_LIST_BY_NAME';
export const REMOVE_SINGLE_COMPARISON_CRITERIA =
  'REMOVE_SINGLE_COMPARISON_CRITERIA';
export const CLEAR_COMPARISON_CRITERIA = 'CLEAR_COMPARISON_CRITERIA';
export const FILTER_PLANET_LIST_BY_COMPARISON =
  'FILTER_PLANET_LIST_BY_COMPARISON';
