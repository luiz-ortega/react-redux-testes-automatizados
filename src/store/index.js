// istanbul ignore file

import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';

const composeSetup =
  process.env.NODE_ENV !== 'production' &&
  typeof window === 'object' &&
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    : compose;

const store = createStore(rootReducer, composeSetup(applyMiddleware(thunk)));

export default store;
