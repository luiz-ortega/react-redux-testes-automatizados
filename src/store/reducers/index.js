import { combineReducers } from 'redux';
import planetReducer from './planets-reducers';

export default combineReducers({ planets: planetReducer });
