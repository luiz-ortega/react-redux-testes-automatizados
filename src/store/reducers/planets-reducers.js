import produce, { setAutoFreeze } from 'immer';

import {
  CLEAR_COMPARISON_CRITERIA,
  FETCH_PLANETS,
  FETCH_PLANETS_FAILURE,
  FETCH_PLANETS_SUCCESS,
  FILTER_PLANET_LIST_BY_COMPARISON,
  FILTER_PLANET_LIST_BY_NAME,
  REMOVE_SINGLE_COMPARISON_CRITERIA,
  SORT_PLANET_LIST,
} from '../types';

/**
 * Immer's auto freeze capabilities don't
 * play well with reselect lib, which we're
 * using in order to implement derived data
 */
setAutoFreeze(false);

export const INITIAL_STATE = {
  list: [],
  loading: false,
  error: null,
  filters: {
    order: {
      column: 'name',
      sort: 'ASC',
    },
    filterByName: {
      name: '',
    },
    filterByNumericValues: [],
  },
};

const reducer = produce((nextState, action) => {
  switch (action.type) {
    case FETCH_PLANETS:
      nextState.loading = true;
      break;
    case FETCH_PLANETS_SUCCESS:
      nextState.loading = false;
      nextState.list = action.planets;
      break;
    case FETCH_PLANETS_FAILURE:
      nextState.loading = false;
      nextState.list = [];
      nextState.error = action.error;
      break;
    case SORT_PLANET_LIST:
      nextState.filters.order.column = action.column;
      nextState.filters.order.sort = action.sort;
      break;
    case FILTER_PLANET_LIST_BY_NAME:
      nextState.filters.filterByName.name = action.name;
      break;
    case FILTER_PLANET_LIST_BY_COMPARISON:
      const { type, ...rest } = action;
      nextState.filters.filterByNumericValues.push({ ...rest });
      break;
    case REMOVE_SINGLE_COMPARISON_CRITERIA:
      const { item } = action;
      const toRemove = Object.values(item).join('-');
      nextState.filters.filterByNumericValues = nextState.filters.filterByNumericValues.filter(
        current => toRemove !== Object.values(current).join('-')
      );
      break;
    case CLEAR_COMPARISON_CRITERIA:
      nextState.filters.filterByNumericValues = [];
      break;
    // no default
  }
}, INITIAL_STATE);

export default reducer;
