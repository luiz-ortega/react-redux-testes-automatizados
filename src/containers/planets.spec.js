import React from 'react';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { render, screen } from '@testing-library/react';
import thunk from 'redux-thunk';
import Planets from './planets';
import { INITIAL_STATE } from '../store/reducers/planets-reducers';
import { FETCH_PLANETS } from '../store/types';

const mockStore = configureMockStore([thunk]);

function renderComponent(overrides = {}) {
  const store = mockStore({
    planets: {
      ...INITIAL_STATE,
      ...overrides,
    },
  });

  const wrapper = render(
    <Provider store={store}>
      <Planets />
    </Provider>
  );

  return { wrapper, store };
}

describe('Component: Planets', () => {
  describe('Displays all elements', () => {
    it('should match snapshot', () => {
      const {
        wrapper: { baseElement },
      } = renderComponent();
      expect(baseElement).toMatchSnapshot();
    });

    it('should render a heading with content Loading...', () => {
      renderComponent({ loading: true });

      const loading = screen.getByRole('heading', { name: /^loading...$/i });

      expect(loading).toBeInTheDocument();
    });

    it('should render a heading with content So sorry...', () => {
      renderComponent({ error: {} });

      const sorry = screen.getByRole('heading', { name: /^so sorry...$/i });

      expect(sorry).toBeInTheDocument();
    });
  });

  describe('Behavior', () => {
    it('should dispatch fetchPlanets action', async () => {
      const { store } = renderComponent();
      const [{ type }] = store.getActions();

      expect(type).toEqual(FETCH_PLANETS);
    });
  });
});
