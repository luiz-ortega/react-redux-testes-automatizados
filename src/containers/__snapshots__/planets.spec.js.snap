// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Planets Displays all elements should match snapshot 1`] = `
<body>
  <div>
    <div
      class="mb-5"
    >
      <div
        class="bg-gray-100 bg-opacity-80 shadow sm:rounded-lg grid grid-flow-row lg:grid-flow-col"
      >
        <div>
          <div
            class="px-4 py-4 text-gray-900"
          >
            <label
              class="font-bold block text-sm uppercase ml-1 text-gray-700"
              for="name"
            >
              Search
            </label>
            <div
              class="mt-1"
            >
              <input
                autocomplete="off"
                class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"
                data-testid="name-filter"
                id="name"
                name="name"
                placeholder="Type the planet's name"
                type="search"
              />
            </div>
          </div>
          <form
            class="px-4 pb-5 text-gray-900"
            data-testid="sorter-form"
          >
            <label
              class="block text-sm font-bold uppercase ml-1 text-gray-700"
              for="Sorting"
            >
              Sorting
            </label>
            <select
              class="capitalize mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
              data-testid="column-sort"
              id=""
              name=""
            >
              <option
                value="name"
              >
                Name
              </option>
              <option
                value="climate"
              >
                Climate
              </option>
              <option
                value="diameter"
              >
                Diameter
              </option>
              <option
                value="orbital_period"
              >
                Orbital Period
              </option>
              <option
                value="gravity"
              >
                Gravity
              </option>
              <option
                value="population"
              >
                Population
              </option>
            </select>
            <div
              class="my-3"
            >
              <label
                class="pr-3"
                for="asc"
              >
                <input
                  checked=""
                  data-testid="column-sort-input"
                  id="asc"
                  name="sortOrder"
                  type="radio"
                  value="ASC"
                />
                 
                ASC
              </label>
              <label
                for="desc"
              >
                <input
                  data-testid="column-sort-input"
                  id="desc"
                  name="sortOrder"
                  type="radio"
                  value="DESC"
                />
                 
                DESC
              </label>
            </div>
            <button
              class="inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
              data-testid="column-sort-button"
              type="submit"
            >
              Apply
            </button>
          </form>
        </div>
        <form
          class="px-4 py-5 sm:p-6 text-gray-900"
          data-testid="filter-numeric-value-form"
        >
          <p
            class="block text-sm font-bold uppercase ml-1 text-gray-700"
          >
            Comparison filter
          </p>
          <div
            class="grid grid-flow-row md:grid-flow-col gap-x-2 mb-2"
          >
            <select
              class="capitalize mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
              data-testid="column-filter"
              id="Columns"
              name="Columns"
            >
              <option
                value="surface_water"
              >
                Surface Water
              </option>
              <option
                value="diameter"
              >
                Diameter
              </option>
              <option
                value="orbital_period"
              >
                Orbital Period
              </option>
              <option
                value="rotation_period"
              >
                Rotation Period
              </option>
              <option
                value="population"
              >
                Population
              </option>
            </select>
            <select
              class="capitalize mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
              data-testid="comparison-filter"
              id="Comparison"
              name="Comparison"
            >
              <option
                value="greater_than"
              >
                Greater than
              </option>
              <option
                value="less_than"
              >
                Less than
              </option>
              <option
                value="equals"
              >
                Equals
              </option>
            </select>
          </div>
          <input
            autocomplete="off"
            class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"
            data-testid="value-filter"
            id="value"
            name="value"
            placeholder="Type the information to compare"
            type="number"
            value="0"
          />
          <button
            class="mt-2 inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
            data-testid="button-filter"
            type="submit"
          >
            Apply
          </button>
        </form>
        <div
          class="px-4 py-5 sm:p-6 text-gray-900 lg:w-96"
        >
          <ul
            data-testid="numeric-filter-selected-data"
          />
        </div>
      </div>
    </div>
    <ul
      class="grid grid-cols-1 gap-6 lg:grid-cols-2"
    />
  </div>
</body>
`;
